### PR TITLE
Recursive descent parser

### PR DESCRIPTION
**Pull request by sn6uv**
## Fixes:
 - #365
 - #260 (but not #381)
 - `Optional`/`Pattern` parsing
 - many other minor parser bugs
 - no more 15mins of silence when rebuilding the tables (when upgrading ply)

## Benefits
* `O(lines)` runtime for multi-line input, previously `O(lines^2)` with 'add a line and retry' approach.
* Parser does not rely on Mathics core so it can be reused for other projects / other purposes by defining additional AST converters.
* Better performance (see below)
* Better syntax errors (see below)

## Downsides
* More complex code (800 lines net increase but about 500 is tests)
* Possible bugs in the immature parser (I think there are actually fewer net bugs)
* More complex `LineFeed` interface for the parser
* We now have both syntax trees and expression trees

## Dependencies
Drops `ply` as a requirement.

## Benchmarks
| test | `master` | `parser` | MMA
|------|------------|------------|------ 
`1+2+...+300` | 11.8ms | 8.01ms |
`1;2;...;300` | 10.5ms | 7.47ms |
`1/2/.../300` | 24.2ms | 14.9ms |
`1^2^...^300` | 16.9ms | 11.0ms |
`! ! ... ! expr` | 10.4ms | 6.63ms |
` expr & & ... &` | 11.1ms | 6.03ms |
` Sin[Sin[...]]` | 29.6ms | 15.1ms |
``Needs["Combinatorica`"]`` | 34.4s | 3.81s | 0.243s
Launch kernel | 2.20s | 1.97s | 0.450s

Performance is marginally better but the main gains are for multi-line input. 

## Better syntax errors
Syntax errors now report more useful information including the line number.
```
$> cat hello.m
Print["Hello World!"]

x * * y
$> mathics -script hello.m
Hello World!
 : "x * " cannot be followed by "* y" (line 3 of "hello.m").
 ```

```
$> mathics -q
In[1]:= Sin[
        
 : Incomplete expression; more input is needed (line 1 of "<stdin>").
In[1]:= x * * y
 : "x * " cannot be followed by "* y" (line 1 of "<stdin>").
In[1]:= 
```

### CODE REVIEW

File: test/test_parser/test_parser.py [hunk 1]
```diff
@@ -0,0 +1,665 @@
...
+        self.check('a ^ b ^ c', 'a ^ (b ^ c)')
+
+    def test_flat(self):
+        self.check('a + b + c', 'Plus[a, b, c]')
+
+    def test_left(self):
+        self.check('a /; b /; c', '(a /; b) /; c')
+
+    def test_Subtract(self):
+        self.check('a - b - c', 'Plus[a, Times[-1, b], Times[-1, c]]')
```
Comment (wolfv) [inline test/test_parser/test_parser.py]: Just curious as to why `-` is parsed as plus-times-minus :)

Comment (sn6uv) [inline test/test_parser/test_parser.py]: That's a good question. The behaviour is correct, see the [wolfram language spec](https://reference.wolfram.com/language/tutorial/OperatorInputForms.html), and it's also what MMA does.

As to why this is part of the language I'm not completely sure but I have some ideas:
* Converting `Subtract[a, b]` to `Plus[a, Times[-1, b]]` saves implementing `Subtract.
* `Plus` and `Times` are flat operators and can, in principle, be multi-threaded. 
* Doing the conversion at parse time prevents the AST from getting too tall (reducing call stack depths when walking the tree).

For interest, something similar happens with `Divide`; `a / b` is converted to `Times[a, Power[b, -1]]`.

Comment (wolfv) [inline test/test_parser/test_parser.py]: Interesting stuff :) Thanks!

**Comment by sn6uv** [comment]
I'm writing up some docs on this (so we know what's going on in future). I think functionality-wise this is finished though. Since this is a pretty major change I want to do a bit more testing before merging though.

**Comment by poke1024** [comment]
very cool :thumbsup:

**Pull request merged**